src/script/conversation/ConversationVerificationStateHandler/shared/conversation/index.ts

```diff
@@ -0,0 +1,139 @@
+    .filteredConversations()
+    .map((conversationEntity: Conversation) => {
+      if (!conversationEntity.removed_from_conversation()) {
+        const userIdsInConversation = conversationEntity.participating_user_ids().concat(userState.self().qualifiedId);
+        const matchingUserIds = userIdsInConversation.filter(userIdInConversation =>
+          userIds.find(userId => matchQualifiedIds(userId, userIdInConversation)),
+        );
+
+        if (!!matchingUserIds.length) {
+          return {conversationEntity, userIds: matchingUserIds};
+        }
+      }
+      return undefined;
+    })
+    .flatMap(activeConversationInfo => (!!activeConversationInfo ? [activeConversationInfo] : []));
+};
+
+interface GetConversationByGroupIdParams {
+  conversationState: ConversationState;
+  groupId: string;
+}
+export const getConversationByGroupId = ({
+  conversationState,
+  groupId,
+}: GetConversationByGroupIdParams): MLSConversation | undefined => {
+  const conversation = conversationState.filteredConversations().find(conversation => conversation.groupId === groupId);
+  if (conversation && isMLSConversation(conversation)) {
+    return conversation;
+  }
+  return undefined;
+};
+
+/**
+ * Check whether to degrade conversation and set corresponding state.
+ *
+ * @param conversationEntity Conversation entity to evaluate
+ * @param shouldShowDegradationWarning Should a modal warn about the degradation?
+ * @returns `true` if conversation state changed to degraded
+ */
+interface WillChangeToDegradedParams {
+  conversationEntity: Conversation;
+  shouldShowDegradationWarning?: boolean;
+  logger: Logger;
+}
+export const willChangeToDegraded = ({
+  conversationEntity,
+  logger,
+  shouldShowDegradationWarning = true,
+}: WillChangeToDegradedParams): boolean => {
+  const state = conversationEntity.verification_state();
+  const isDegraded = state === ConversationVerificationState.DEGRADED;
+
+  if (isDegraded) {
+    return false;
+  }
+
+  // Explicit Boolean check to prevent state changes on undefined
+  const isStateVerified = state === ConversationVerificationState.VERIFIED;
+  const isConversationUnverified = conversationEntity.is_verified() === false;
+  if (isStateVerified && isConversationUnverified) {
+    conversationEntity.verification_state(
+      shouldShowDegradationWarning ? ConversationVerificationState.DEGRADED : ConversationVerificationState.UNVERIFIED,
+    );
+    logger.log(`Verification of conversation '${conversationEntity.id}' changed to degraded`);
+    return true;
+  }
+
+  return false;
+};
+
+/**
+ * Check whether to verify conversation and set corresponding state
+ *
+ * @param conversationEntity Conversation entity to evaluate
+ * @returns `true` if conversation state changed to verified
+ */
+interface WillChangeToVerifiedParams {
+  conversationEntity: Conversation;
+  logger: Logger;
+}
+export const willChangeToVerified = ({conversationEntity, logger}: WillChangeToVerifiedParams): boolean => {
```

Comment:
Naming is a little weird here `willChangeToVerified` seems to imply that it will just return a boolean. 
But it's actually also changing the verified state of the conversation. 

We have a single responsibility principe failure here. Not sure how this could be handled better though 🤔

Reply:
Lets go with "attemptChangeToVerified" and "attemptChangeToDegraded"

Reply:
yeah, at least it conveys the right idea. I'm fine with it 👌